addons/barcodes/models/barcode_rule.py

```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import re
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class BarcodeRule(models.Model):
+    _name = 'barcode.rule'
+    _order = 'sequence asc'
+
+    name = fields.Char('Rule Name', required=True, help='An internal identification for this barcode nomenclature rule')
+    barcode_nomenclature_id = fields.Many2one('barcode.nomenclature', 'Barcode Nomenclature')
+    sequence = fields.Integer('Sequence', help='Used to order rules such that rules with a smaller sequence match first')
+    encoding = fields.Selection('_encoding_selection_list', 'Encoding', required=True, default='any', help='This rule will apply only if the barcode is encoded with the specified encoding')
+    rule_type = fields.Selection('_get_type_selection', 'Type', required=True, default='product', oldname='type')
```

Comment:
can you make separate commit regarding changes of field name like type ==> rule_type